synapse/logging/opentracing.py

```diff
@@ -930,20 +930,29 @@ def call_back(result: R) -> R:
                         return result
 
                     def err_back(result: R) -> R:
+                        # XXX: Feels like we could put the error details into the
+                        # `scope.__exit__(...)`
                         scope.__exit__(None, None, None)
                         return result
 
                     result.addCallbacks(call_back, err_back)
 
+                elif inspect.isawaitable(result):
+                    # TODO: I don't know how to type this
+                    async def await_coroutine():
+                        try:
+                            awaited_result = await result
+                            scope.__exit__(None, None, None)
+                            return awaited_result
+                        except Exception as e:
+                            scope.__exit__(type(e), None, e.__traceback__)
+                            raise
+
+                    # The original method returned a coroutine, so we create another
+                    # coroutine wrapping it, that calls `scope.__exit__(...)`.
+                    return await_coroutine()
```

Comment:
Need some help typing this and double-checking if this is correct.

Reply:
I also have no idea how to type this...

A not very elegant work around might be to move this check before the `Deferred` check?

```python
if inspect.isawaitable(result):
    result = defer.ensureDeferred(result)

if isinstance(result, defer.Deferred):
    ...
```

Reply:
It's pretty clever! Seems good enough to me and less boilerplate anyway

Great shout!

Reply:
Actually seeing a type-check lint from trying to use this:
```
synapse/logging/opentracing.py:933: error: Argument 1 to "ensureDeferred" has incompatible type "Awaitable[Any]"; expected "Union[Coroutine[Deferred[<nothing>], Any, <nothing>], Generator[Deferred[<nothing>], Any, <nothing>], Deferred[<nothing>]]"  [arg-type]
```

And I can't figure out how to generically from `Awaitable` to `Coroutine` without the same problem of not knowing how to type it.

Reply:
Ugh, maybe I'm wrong you can't use that with `awaitables` at all, that would be quite annoying.

Let me poke at it a bit?

Reply:
I think what we do for `_AsyncResource._async_render` is clearer:

https://github.com/matrix-org/synapse/blob/9c96b06cd2656ca46d17de17571591f08a2a83ca/synapse/http/server.py#L345-L353

Reply:
> We don't have the `async` function context to just do that directly which is why I've been using the cumbersome wrapper.

Yeah my suggestion was to always just do it inside an async function, if you can await you await, otherwise you just return.

The downside of this is that the output of the wrapper will always be async, which I don't think happens right now? (This is similar to what we do with `cached` where the wrapper always wraps into a `Deferred`.)

> And I don't think that solves the return type problem since `R` and the result of `await R` are different things :thinking: 

We get around that in `_AsyncResource` with various types of `type: ignore[...]` I think. Or it is all just `Any` and type hints aren't helping us here.

Reply:
> Or it is all just `Any` and type hints aren't helping us here.

Ahh, I think that's the key! I've reverted back to the original wrapped coroutine `scope.__exit__(...)` implementation but updated the `def _wrapper(...) -> Any`.

This matches what we were already doing for the `async` function logic branch anyway as well.


> The downside of this is that the output of the wrapper will always be async, which I don't think happens right now?

I've opted not to do this for now.